Feat: Add Jina AI Embeddings support

Harsha-Karimikonda opened the pull request:
feat: Add Jina AI Embeddings Integration

Adds `JinaEmbeddings` class supporting Jina AI's embedding API.

**Key Features:**
- Single/batch embedding generation
- Token counting and text similarity
- API error handling with retries
- Environment variable or parameter-based auth

**Test Coverage:**
- Unit tests for all public methods
- Integration tests (skipped without `JINA_API_KEY`)

**Usage:**
```python
from chonkie.embeddings import JinaEmbeddings
embeddings = JinaEmbeddings()  # Uses JINA_API_KEY

chonknick commented:
Hey @Harsha-Karimikonda! 👋

Thanks for opening a PR! We've been wanting to add JinaEmbeddings support for a while now, this is really cool!

I'll review the PR sometime today or tomorrow and once everything looks good, we'll go ahead and merge it. 🥳

Thankss 😊

chonknick commented on src/chonkie/embeddings/jina.py:
```diff
@@ -0,0 +1,267 @@
+        )
+
+    def count_tokens(self, text: str, tokenizer: str = 'cl100k_base') -> int:
+        """Count tokens in text using the Jina segmenter.
+
+        Args:
+            text (str): The input text.
+            tokenizer (str): The tokenizer model to use.
+
+        Returns:
+            int: The number of tokens in the text.
+
+        Raises:
+            requests.exceptions.RequestException: If the API request fails after retries.
+
+        """
+        if not text:
+            return 0
+            
+        url = 'https://api.jina.ai/v1/segment'
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': f'Bearer {self.api_key}'
+        }
+
+        data = {
+            "content": text,
+            "tokenizer": tokenizer
+        }
+        
+        for attempt in range(self._max_retries):
+            try:
+                response = requests.post(url, headers=headers, json=data)
+                response.raise_for_status()
+                return response.json()['num_tokens']
+            except requests.exceptions.RequestException as e:
+                if attempt == self._max_retries - 1:
+                    raise
+                warnings.warn(f"Attempt {attempt + 1} failed: {str(e)}. Retrying...")
+
```
Hey @Harsha-Karimikonda!

The segmenter API is doesn't actually count the tokens with the jina embedding models, it does so with OpenAI tokenizers while JinaAI's embedding models use separate tokenizers. Check: [https://huggingface.co/jinaai/jina-embeddings-v3](https://huggingface.co/jinaai/jina-embeddings-v3)

You'd be better off loading a `tokenizers.Tokenizer` object in the `__init__` and then using that to count tokens.

chonknick commented on src/chonkie/embeddings/jina.py:
```diff
@@ -0,0 +1,267 @@
+        """
+        return self._dimension
+        
+    def get_tokenizer_or_token_counter(self):
+        """Get the tokenizer or token counter for the embeddings.
+
+        Returns:
+            Callable[[str], int]: The method used for counting tokens.
+
+        """
+        return self.count_tokens
```
Hey @Harsha-Karimikonda!

It would be better to pass the tokenizer object directly here once you load the tokenizer in the `__init__`, as mentioned in the previous review comment.

chonknick commented on tests/embeddings/test_jina_embeddings.py:
```diff
@@ -0,0 +1,193 @@
+import numpy as np
+import pytest
+
+# Ensure the src directory is in the path for imports
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../src')))
```
Hey @Harsha-Karimikonda, 

Could you explain what this line does and why it's needed here? 

Thanks 😊

Harsha-Karimikonda replied:
Hey @chonknick ,

This is my system specific issue, apologies for oversight. Sometimes it throws a "module not found error" when i try to import from other files. This line is removed now.

chonknick commented on tests/embeddings/test_jina_embeddings.py:
```diff
@@ -0,0 +1,193 @@
+    assert isinstance(embedding_model.dimension, int)
+    assert embedding_model.dimension == 1024 # Check default dimension
+
+@skip_if_no_key
+def test_get_tokenizer_or_token_counter(embedding_model):
+    """Test the get_tokenizer_or_token_counter method.
+
+    Args:
+        embedding_model: The JinaEmbeddings instance fixture.
+
+    """
+    counter_func = embedding_model.get_tokenizer_or_token_counter()
+    assert callable(counter_func)
+    # Check if it returns the count_tokens method
+    assert counter_func.__func__ is JinaEmbeddings.count_tokens
```
Hey @Harsha-Karimikonda, 

The test for this should also match the impl mentioned previously and should check for a `tokenizers.Tokenizer` object.

chonknick commented on src/chonkie/embeddings/jina.py:
```diff
@@ -0,0 +1,267 @@
+            except requests.exceptions.HTTPError as e:
+                if len(batch) > 1:
+                    warnings.warn(f"Batch embedding failed: {str(e)}. Trying one by one...")
+                    print(f"Fallback to single embeddings due to: {str(e)}")
```
Hey @Harsha-Karimikonda!

The statement for fallback should also be a `warnings.warn` instead of a `print`. Could you change this? 

Thanks 😊

chonknick commented on src/chonkie/embeddings/jina.py:
```diff
@@ -0,0 +1,267 @@
+            task: str = "text-matching",
+            late_chunking: bool = True,
+            embedding_type: str = "float",
+            dimensions: int = 1024,
```
Hey @Harsha-Karimikonda!

The `dimensions` shouldn't be a user input but instead should be picked up based on the model passed in. Ideally, we would probably send a dummy request to the API if we don't know the dimensions already. 

We can also make a simple Dictionary with all the models from JinaAI and their dimensions so that we don't have to send a dummy request for each of them — and if we don't have a model in the dictionary, we can fallback to the dummy request to infer the dimenstions. 

Thanks! 😊

chonknick reviewed (changes requested):
Hey @Harsha-Karimikonda, 

Just left a couple of points of feedback, the main ones being the `dimensions` should be inferred and the `tokenizer` object to be created. 

Once those changes are made, I believe we can merge this PR. 

Thanks 😊

Harsha-Karimikonda commented:
Hey @chonknick 

I have made changes accordingly, please review it and let me know if any further improvements needed.

Changes made:
Removed Segmenter API usage and added AutoTokenizer.
Made changes accordingly to test cases.
Added a available models dict to get the dimensions

chonknick commented:
Hey @Harsha-Karimikonda! 👋

Thanks for the changes, LGTM~ Merging... 

Welcome to Chonkie's Contributors fam 🤗🎉

Thanks 😄
